### PR TITLE
[WIP] Hooks OOP design

### PR DESCRIPTION
**Pull request by chsasank**
Instead of using hooks as a dict, we can borrow from OOP design of Keras Callbacks.
We can now extend the `Hook` object to create learning rate schedulers, etc [like in keras](https://keras.io/callbacks/).

I also made a few changes and added comments to make the engine code more readable. I will explain more in inline comments.

Todo:

- [x] Create a list hook object like [here](https://github.com/fchollet/keras/blob/master/keras/callbacks.py#L28)
- [ ] Creare LRSchedulers, model checkpoints etc.

Sasank,

### CODE REVIEW

File: torchnet/engine/engine.py [hunk 1]
```diff
@@ -1,71 +1,87 @@
...
                 state['sample'] = sample
-                self.hook('on_sample', state)
+                self.hook.on_sample(state)
 
                 def closure():
+                    # forward
                     loss, output = state['network'](state['sample'])
                     state['output'] = output
                     state['loss'] = loss
+                    self.hook.on_forward(state)
```
Comment (chsasank) [inline torchnet/engine/engine.py]: making loss.backward() after `on_forward`

File: torchnet/engine/engine.py [hunk 2]
```diff
@@ -1,71 +1,87 @@
...
+
                     return loss
 
                 state['optimizer'].zero_grad()
                 state['optimizer'].step(closure)
                 state['t'] += 1
+
+            self.hook.on_end_epoch(state)
             state['epoch'] += 1
-            self.hook('on_end_epoch', state)
```
Comment (chsasank) [inline torchnet/engine/engine.py]: previously, epoch number increasing before `on_end_epoch`

File: torchnet/engine/hook.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class Hook(object):
+    def __init__(self):
+        pass
```
Comment (chsasank) [inline torchnet/engine/hook.py]: Here, we have a choice to take a dict like {'on_start': function}

**Comment by szagoruyko** [comment]
@chsasank any progress on this, what do you plan to add next? there seems to be a collision now with LR scheduling in the core https://github.com/pytorch/pytorch/pull/1370
